File: packages/sdk/sdk-py/src/phaseo/testing.py
"""Strict fixture transport for local application tests, with no network fallback."""
from __future__ import annotations

from typing import Any
import httpx


class MockTransport(httpx.MockTransport):
    def __init__(self, fixtures: list[dict[str, Any]]):
        self.fixtures = list(fixtures)
        self.requests: list[httpx.Request] = []
        super().__init__(self._handle)

    def _handle(self, request: httpx.Request) -> httpx.Response:
        if not self.fixtures:
            raise AssertionError(f"Unexpected mock request: {request.method} {request.url.path}")
        fixture = self.fixtures.pop(0)
        assert request.method == fixture["method"].upper(), "Mock HTTP method mismatch"
        assert request.url.raw_path.decode().split("?", 1)[0] == fixture["path"], "Mock HTTP path mismatch"
        self.requests.append(request)
        payload = {"content": fixture["body"]} if "body" in fixture else {"json": fixture["json"]} if "json" in fixture else {"content": b""}
        return httpx.Response(fixture.get("status", 200), headers=fixture.get("headers"), **payload)

    def assert_done(self) -> None:
        assert not self.fixtures, f"{len(self.fixtures)} mock requests unused"


def job_fixtures(path: str, job_id: str, outcome: str = "completed") -> list[dict[str, Any]]:
    from urllib.parse import quote
    return [
        {"method": "POST", "path": path, "json": {"id": job_id, "status": "queued"}},
        {"method": "GET", "path": f"{path}/{quote(job_id, safe='')}", "json": {"id": job_id, "status": "running"}},
        {"method": "GET", "path": f"{path}/{quote(job_id, safe='')}", "json": {"id": job_id, "status": outcome}},
    ]

File: packages/sdk/sdk-ts/src/testing.ts
export type Fixture = {
  method: string;
  path: string;
  status?: number;
  headers?: Record<string, string>;
  json?: unknown;
  body?: string;
};

/** Strict, ordered fixtures. Unexpected calls fail locally and never reach a provider. */
export function createMockTransport(fixtures: readonly Fixture[]) {
  let position = 0;
  const requests: Array<{ method: string; url: string; body: unknown }> = [];
  const fetchImpl: typeof fetch = async (input, init) => {
    const request = new Request(input, init);
    request.signal.throwIfAborted();
    const fixture = fixtures[position++];
    const path = new URL(request.url).pathname;
    if (!fixture || fixture.method.toUpperCase() !== request.method || fixture.path !== path) throw new Error(`Unexpected mock request: ${request.method} ${path}`);
    const raw = await request.text();
    let body: unknown = raw;
    try { body = JSON.parse(raw); } catch { /* Binary/plain text requests stay strings. */ }
    requests.push({ method: request.method, url: request.url, body });
    return new Response(fixture.body ?? (fixture.json === undefined ? null : JSON.stringify(fixture.json)), {
      status: fixture.status ?? 200,
      headers: { ...(fixture.json === undefined ? {} : { "content-type": "application/json" }), ...fixture.headers },
    });
  };
  return { fetchImpl, requests, assertDone() { if (position !== fixtures.length) throw new Error(`${fixtures.length - position} mock requests unused`); } };
}

export function jobFixtures(path: string, id: string, outcome: "completed" | "failed" = "completed"): Fixture[] {
  return [
    { method: "POST", path, json: { id, status: "queued" } },
    { method: "GET", path: `${path}/${encodeURIComponent(id)}`, json: { id, status: "running" } },
    { method: "GET", path: `${path}/${encodeURIComponent(id)}`, json: { id, status: outcome } },
  ];
}
